Enhance: enable builds in a full Windows MSYS2 environment

SlySven opened the pull request:
By defining `WITH_MAIN_BUILD_SYSTEM` to the value `NO` this PR makes enough changes to the qmake project file to enable Mudlet to be compiled in a full MSYS2 development environment (in the MSYS2 Qt Creator) - this will enable easier development by Windows users (particularly those who also have some familiarity with *nix systems) as I have documented at:
**[Compiling on Windows 7+ (MSYS2_Alternative)](https://wiki.mudlet.org/w/Compiling_Mudlet#Compiling_on_Windows_7.2B_.28MSYS2_Alternative.29)**.

It also makes some changes to the setting up of the LUA package paths for the lua code formatter so that the paths are all entered with Unix style directory separators but converted to whatever the Lua package handler is set to use. In a Windows environment it is not unheard of to get both '\\' and '/' being used within the same path as different parts get generated in stages - and using the backslash one can produce surprising error messages if the back slash is not properly escaped when displaying those messages in the main console or elsewhere (they disappear entirely or end up escaping following characters producing misleading information).

*Edit: although intended originally just for QMake on Windows - a need to consider an additional library (for a crash handler) that only had a CMake build system prompted me to also consider and try to fix things to handle the CMake case - which had not been used for Windows for some years - as well...*

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Enhance: enable builds in a full Windows MSYS2 environment (QMake only)

By defining `WITH_MAIN_BUILD_SYSTEM` to the value `NO` this PR makes enough
changes to the qmake project file to enable Mudlet to be compiled in a full
MSYS2 development environment (in the MSYS2 Qt Creator) - this will enable
easier development by Windows users (particularly those who also have some
familiarity with *nix systems) as I have documented at:
"Compiling on Windows 7+ (MSYS2_Alternative)" but it seems that the URL is
causing GitHub to forget about the PR as it seems to push and is recorded
in my local repository but never actually lands there!

It also makes some changes to the setting up of the LUA package paths for
the lua code formatter so that the paths are all entered with Unix style
directory separators but converted to whatever the Lua package handler is
set to use. In a Windows environment it is not unheard of to get both '\'
and '/' being used within the same path as different parts get generated
in stages - and using the backslash one can produce surprising error
messages if the back slash is not properly escaped when displaying those
messages in the main console or elsewhere (they dissappear entirely or
end up escaping following characters producing misleading information).

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into windows_development

Needed to import fix for CI build failures on Travis MacOs

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

keneanung commented:
/refresh links

vadi2 commented:
None of us on the team know MSYS2 and I don't see and willpower to learn it either in order to do support and bugfixing - if we put it in, I think it should be unofficial like FreeBSD.

SlySven commented on src/mudlet.pro:
```diff
@@ -271,23 +283,51 @@ unix:!macx {
+
+        INCLUDEPATH += \
+             "C:\\Libraries\\boost_1_71_0" \
+             "$${MINGW_BASE_DIR_TEST}\\includ"e \
```
Ooops! A typo that ought to break things:
```suggestion
             "$${MINGW_BASE_DIR_TEST}\\include" \
```

SlySven committed:
Update src/mudlet.pro

BugFix: fix a typo in qmake project file.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
Well you are already installing and using **Mingw/MSYS** things but that is a little suspect in IMHO as they just do not offer a 64-bit path in the future and actually could clash with the limited **Mingw-w64** (32-bit parts) that Qt *does* ship in their Windows Qt SDK!

What this PR does is allow some things to be changed so that qmake will work in either the existing case or a full MSYS2/Mingw-w64 one - so the CI process here is actually checking that I haven't balls-ed up anything for the existing case.  The other part of this PR revises the path handling - particularly for the external Lua stuff so that the requisite files can be found at run-time. For instance, until this PR I realised that the Lua code formatter may not work for MacOs builds of Mudlet in a development environment - as AFAICT it ONLY looked in the resource bundle for it - which as someone hacking the source code on that platform is not going to have...!

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -17161,46 +17161,140 @@ void TLuaInterpreter::initIndenterGlobals()
+     */
+#if defined(LUA_DEFAULT_PATH)
+    if (!QStringLiteral(LUA_DEFAULT_PATH).isEmpty()) {
+        luaL_dostring(pIndenterState.get(), QStringLiteral("package.path = [["
```
This seems to be duplicated except it prepends LUA_DEFAULT_PATH if it exists - can be written better.

SlySven replied:
> This seems to be duplicated except it prepends LUA_DEFAULT_PATH if it exists ...

Sort of, it prepends `LUA_DEFAULT_PATH` if it *exists* **and is not an empty string**.

> ... can be written better.

How so?

vadi2 replied:
Yes just append the string if it's there.

I've looked the code and the function is now absolutely massive with a tangle of #ifdefs that make the logic really hard to follow and huge comments you have to read. That is not maintainable, we can't have this - we need to have code that everyone finds nice & pleasant to work with.

SlySven replied:
>Yes just append the string if it's there.

That won't work, if the variable is defined but is *an empty string* then it will cause the path to have `/?.lua` included which means look for it in the root directory - on most OSes that location is readable but not writeable for *normal* users. If you refer back to #1616 you will note part of the problems there was that `LUA_DEFAULT_PATH` was getting defined but not set to anything - and that is very hard to detect in the C/C++ preprocessor and cannot be done cleanly.

vadi2 replied:
This check can be done at runtime as far as I can see just fine and we'll have much simpler code, no?

SlySven replied:
Well it eliminated some preprocessor conditional lines and still *seems* to work so: I guess so.

vadi2 reviewed (changes requested):
`initIndenterGlobals()` was ~60 lines when I added it, and it is now **2x the size** without much gained functionality. The amount of commenting you're adding is not OK - gotta be considerate of others using this code too.

Please trim the comments. They are not helping, they are hurting.

SlySven committed:
Revise: reduce almost duplicate comments

Fix an addition to the `package.cpath` which would not have worked as it
did not specify the file extension which is OS dependent.

Also remove LuaJIT remenent, which we dropped support for a long time back.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
I've tried to trim down the main big comment and removed the similar duplicate for the Windows case by instead using numbers and appending them to the relevant lines. Is that good enough for something that is *quite* OS dependent.

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -16963,25 +16963,29 @@ void TLuaInterpreter::initLuaGlobals()
+#endif
 
 #if defined(Q_OS_LINUX)
-    // if using LuaJIT, adjust the cpath to look in /usr/lib as well - it doesn't by default
```
Note the removal of the LuaJIT stuff following a recent, brief discussion we had about it on Discord!

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -17158,49 +17162,122 @@ void TLuaInterpreter::initIndenterGlobals()
+     * directory two parent directories up:
+     * 1 installed *nix case - probably not applicable to Windows
+     *     "LUA_DEFAULT_PATH/?.lua" (if defined and not empty)
+     * 2 AppImage (subdirectory of executable):
```
Oops - comment should be *actual directory* of the executable.

SlySven replied:
```suggestion
     * 2 AppImage (directory of executable):
```

SlySven committed:
Apply suggestions from code review

Revise: fix an error in a comment.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Enhance: modifications to get CMake working on Windows

Revise a Mudlet specific CMake macro to not include the word "module" as
that is not appropriate for all usages now.

Fix an obscure CMake build error caused by the use of:
`LIBRARYNAME::LIBRARYNAME` in `target_link_libraries(...)` which is the
form for an interface usage of a library - this causes a failure of the
build with an error message of the form:
`src/CMakeFiles/mudlet.dir/build.make:1954: *** target pattern contains
no '%'.  Stop.` that line is actually one about one of the libraries
concerned - and it is the first one which shows up in that file with a
LIBRARYNAME-NOTFOUND entry. The fix seems to be to use only a LIBRARYNAME
form.

Fix a problem in `(static QString) mudlet::getShortPathName(const QString&
name)` which is cause by a Windows specific function that takes.returns
template/typedef type arguments which only work if the symbols
`UNICODE` and `_UNICODE` to be defined and which aren't in an MSYS2/
Mingw-w64 environment.

Revise some usages of the APP_BUILD defined value so that they are
handled correctly (using `QStringLiteral`/`QByteArray` wrapppers).

Change the CMake find module for Pugixml so that it uses a variable name in
ALL_UPPER_CASE to remove a developer warning about using a mixed case one.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: take out some pre-processor stuff as run-time code works without it

I was a bit sceptical at first but it *seems* to work.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 reviewed (commented):
It's looking better :grinning: @demonnic would you mind reviewing the changes for readability as well?

demonnic commented on src/TLuaInterpreter.cpp:
```diff
@@ -16963,25 +16963,29 @@ void TLuaInterpreter::initLuaGlobals()
 #elif defined(Q_OS_WIN32)
+// The following is only needed in the original powershell script based Mudlet
+// build environment - and should probably have been done via the LUA_CPATH
+// environmental variable:
```
Then why not fix that instead of making it a comment and adding another ifdef?

SlySven replied:
Because I do not want to force changes in other things (a build system using PowerShell which I do not grok) which might break them.

vadi2 replied:
The new comment is wrong, this is needed when you're running Mudlet in WIndows using the setup script. I recommend removing the comment as it's adding little value.

SlySven replied:
> The new comment is wrong, this is needed when you're running Mudlet in WIndows using the setup script.

That is what I was saying and is how the logic is setup - so your remark doesn't compute. I called it "the original powershell script" you call it " in WIndows using the setup script" but we mean the same thing - don't we?

vadi2 replied:
Yes, sorry, we're talking about the same thing - and no it should not be done via `LUA_CPATH`. This setup is just fine as well. This comment needs to go, code is not a place for personal contemplations.

demonnic commented on src/TLuaInterpreter.cpp:
```diff
@@ -17158,49 +17162,115 @@ void TLuaInterpreter::initIndenterGlobals()
+                      .arg(QCoreApplication::applicationDirPath()).toUtf8().constData());
+    } else {
+        luaL_dostring(pIndenterState.get(), QStringLiteral("package.path = [["
+                                                           "%1/?.lua;" // 2
```
all of this section is reproduced code. Instead of two big repeated sections, why not do the push of the LUA_DEFAULT_PATH in the if, and then push the rest of them outside of the if/else and also keep it DRY.

If the existing package.path absolutely must be behind all of it, then create the string first and then adjust package.path. But this huge repeated block is not only an eyesore, it's just begging for someone to miss an update in one or the other of them.

SlySven replied:
Okay I've tried a different approach - is that better now?

demonnic commented on src/TLuaInterpreter.cpp:
```diff
@@ -17158,49 +17162,115 @@ void TLuaInterpreter::initIndenterGlobals()
+     * NOTE: In the cmake case the shadow build directory is alongside the base
+     * directory of the extracted source package so it MUST be extracted into
+     * a known directory name which is assumed here to be "mudlet"!
+     */
```
This is a case of too much information, I think. We don't need to explain how lua require() works in our C++ files complete with example. This beast of a file is already 17k+ lines long, we don't need to make it even harder for our editors to open. 

As for delineating each case, I can see an argument for this but I feel like the right place to do it is above each entry when you add it. Which will be made easier if you take my suggestion to keep the code DRY below.

SlySven replied:
I had it about each line in each case - twice (one for \*nix like cases and once for Windows) - and Vadim said I should cut it down. As for explaining each entry it took a far bit of research/hackery to work out why each line was needed and I don't want that information lost. For instance for CMake *shadow (out-of-source, the default these days)* builds it is ***essential*** to extract the mudlet source into a directory called, not surprisingly, "mudlet" - but that is not obvious!

vadi2 replied:
The middle of a 17,000 long file is definitely not the right place to have this information.

SlySven replied:
Also:
>  We don't need to explain how lua require() works in our C++ files complete with example.

We need to explain it somewhere - it took me a while to workout that the '.'s in that `require` had significance and if I had had that explanation when I first tried porting Mudlet to both MSYS2 AND FreeBSD it would have made things go a lot smoother and without the head scratching around: WTF is `base.lua` and why is it that that file is being sought?! Given that this is the code that is looking for it it seems a perfectly logical place to say what it is looking for. Remember, I am primarily a C/C++ coder not a Lua one.

vadi2 replied:
It's explained in the Lua manual - part of the learning process. No Mudlet dev should have personal comments in the codebase, it's shared by all!

SlySven replied:
There is nothing about that that is a personal comment? Where are you forming that opinion from? :confused:

vadi2 replied:
Well - I don't think others find it helpful, certainly me and demonnic don't. I think you'd be the only one to benefit from it. 

If there's already a proper manual explaining something, we shouldn't duplicate it inside Mudlet.

vadi2 replied:
This needs to go as well, this explanation is not necessary. If you'd like to record for yourself how the loader works, it's best to do it elsewhere - because forcing everyone to read 6 line explanation of how loading of just one particular module we use isn't nice to the team. And it's forced because comments are meant to be read - if we start ignoring comments by habit because they're long and not useful, that will be a bad situation!

SlySven replied:
Some explanation *is* necessary IMHO!

I have shorted it to three lines and included a link to a very informative SE QA - some of the details there (about '.'s being used as directory separators - do not *even* get mentioned in the PIL 5.1 manual) - which is why I felt it necessary to put something in for those of us who are **NOT** Lua gurus...!

vadi2 replied:
I still don't think it's helpful. You'll be looking at the Lua manual anyway to learn it. There is no added value by putting it here, except making everyone read it, and not everyone wants to read this. :confused:

demonnic commented on src/TLuaInterpreter.cpp:
```diff
@@ -17158,49 +17162,115 @@ void TLuaInterpreter::initIndenterGlobals()
+    }
+
+#if ! defined (Q_OS_MACOS)
+    // AppInstaller on Linux would like the search path for the binary modules
```
Do we really have to split this to two lines? it seemed perfectly fine on one line before and I doubt anyone is trying to edit this on an 80 character terminal these days.

SlySven replied:
It is a pain having long lines if one is using a side-by-side split view (useful if comparing similar blocks of code in two different places {ideally to see if there is scope for refactoring to a new function}) as I often do when editing in Qt Creator.

As a guide I tend to enable a (80 column wide) visible margin in the IDE.

demonnic commented on src/TLuaInterpreter.cpp:
```diff
@@ -17158,49 +17162,115 @@ void TLuaInterpreter::initIndenterGlobals()
+end)LUA");
+    // clang-format on
+
+    if (!QStringLiteral(LUA_DEFAULT_PATH).isEmpty()) {
```
Isn't this exactly the same thing as above, apart from the toNativeSeparators function? Why not define the function so it handles both unix and windows variants, and then do this one time using the function? We don't need to make this super performant, it happens one time on startup. Keeping the code DRY would make it MUCH more maintainable and a lot less frustrating to read in future.

SlySven replied:
We don't need it for non-Windows OSes because they do not - as far as I understand it - tend to use anything other than `/`as the second element of `package.config` which is the directory separator that the Lua package handler uses. This PR is trying to ease in handling alternatives for the Windows case but without changing unnecessary things for the other OS. OTOH it might be simpler to have less variations. I will take another look.

SlySven replied:
Ah, no, it does not have the number 2 entry.

demonnic replied:
It's not about 'does unix need it' it's about 'do we need to maintain this in multiple locations' and the answer is 'no, we do not' 

Does having case 2 in the package.path in the below case cause issues? If so, does it need to be in position 2 for everything else? If not, shove it at the end of the string you're assembling before modifying package.path. 

But I'm pretty confident the answer isn't a bunch of otherwise repeated code we have to maintain if something changes.

SlySven replied:
I've had a go at simplifying things now - does it look any better now?

demonnic replied:
Much better, thank you =) 
I still think the comments are perhaps a bit much, but the package.path/cpath stuff no longer sets off that twitch in my eye ;)

SlySven replied:
They will help in the future when some poor soul is trying to work out why the files won't load on their OS - even though they think they have the right files in the right place - I had to jiggle things around to get the LCF working on FreeBSD as I recall...

... and actually the revision has made the code a bit easier to fix if something changes in the future.

As for your eye twitch - remember: "*In the country of the blind, the one-eye man will be king!*" :stuck_out_tongue_winking_eye:

demonnic commented on src/TLuaInterpreter.cpp:
```diff
@@ -17295,7 +17365,7 @@ void TLuaInterpreter::loadGlobal()
 #endif
         if (!(QFileInfo(pathFileName).isReadable())) {
-            failedMessages << tr("%1 (isn't a readable file or symlink to a readable file)", "This is not a file or a symbolic link to a file").arg(pathFileName);
+            failedMessages << tr("%1 (isn't a readable file or symlink to a readable file)", "This is not a file or a symbolic link to a readable (for this user) file").arg(pathFileName);
```
I feel like this is more specificity than is needed and is more likely to confuse the issue than help most people.

SlySven replied:
It is only going to show if there is a problem - and if there is a problem it helps to know as much detail as we can reasonable establish - in this case that there is something there but it isn't a file or symlink to a file we can read...

vadi2 replied:
I don't think so. To be honest "This is not a file or a symbolic link to a readable (for this user) file" makes little sense when you read it.

SlySven replied:
Ha, it is a comment for the translators not a UI text. I suppose I could cut it out entirely and let them work it out from the Engineering text...

Kebap replied:
It basically repeats the first text, not adding any more information

SlySven replied:
Did you mean: "the explanation text for the translators does not add any more information compared to the source text"? Well it does explicitly point out that the problem is likely to be permissions (*read-permissions* for *this* user") - I do not have the source open around this - but - IIRC - there is a prior test that checks for the *existence* of the file prior to this check that the user can read it...

demonnic commented on src/mudlet.pro:
```diff
@@ -204,6 +204,18 @@ isEmpty( OWN_QTKEYCHAIN_TEST ) | !equals( OWN_QTKEYCHAIN_TEST, "NO" ) {
   DEFINES += INCLUDE_OWN_QT5_KEYCHAIN
 }
 
+########## Full MSYS2 (alternative Windows build) environment support ##########
```
You already have a wiki link in here for documentation, perhaps all this extra info could be put on the wiki as well? This could be a link to the wiki and a mention that it can be enabled using `WITH_MAIN_BUILD_SYSTEM="NO"`

SlySven replied:
I don't know about other people but if I am hacking around in an unfamiliar project I want the configuration details to hand in the file I am looking through looking for clues - and not go looking for on-line reference material which has been forgotten. Fair enough that *shouldn't* happen here but it is a lot simpler to document stuff here as I need to and then perhaps migrate it to an external reference (Wiki) when it is a bit more stable... 

... OTOH I will think about it.

SlySven replied:
I have thought about it and my opinion is that the text present says:
* that it is only relevant for Windows developers
* that there is more information in the Wiki
* that it changes some paths and filenames
* that to enable it you need to set an environmental variable

It doesn't strike me that any of that is unhelpful to have there or that it is unduly verbose. What alternative text would you consider acceptable instead?

demonnic replied:
The wiki link clearly states it is windows, and MSYS2. You could just as easily convey this information as 
```
# Windows MSYS2 builds require some altered paths
# see https://wiki.mudlet.org/w/Compiling_Mudlet#Compiling_on_Windows_7.2B_.28MSYS2_Alternative.29
# enable using WITH_MAIN_BUILD_SYSTEM="NO"
```

Half the lines, all the relevant information, and the link to the wiki for further information. You're not necessarily meant to be having a friendly chat over coffee with every future developer who comes across your comments

demonnic reviewed (changes requested):
I feel like there is a lot of repeated code here and this should be refactored to keep it DRY, as it currently stands it's more tech debt than supporting MSYS2 for one or two users is worth.

SlySven committed:
Edit: fix copy paste issue

I thought something needed to be more conditional than it was but didn't
get it undone in last commit.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Refactor: simplify the code to set up the Lua/C additional paths for LCF

Peer-review suggested I needed to shrink the code.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: shrink some comments

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

jfmherokiller commented:
btw we can check for msys2 by using `if(MSYS)` inside of cmake i think

SlySven commented:
> btw we can check for msys2 by using if(MSYS) inside of cmake i think

You mean `if(MSYS2)` don't you?

We have to be a bit cautious as there are some bits that it shares with the project from which it was forked and Qt's own Windows SDK does use some elements that are shared with the MSYS2/Mingw-w64 environment.

Effectively the full MSYS2 setup actually provides three separate environments a bare-bones more \*nix-like "MSYS2" one which is where auto-tool-ised libraries can be built from the command line and two more Windows like "Mingw32" and "Mingw64" environments where GUI applications are built for the corresponding bit-nesses. The Qt SDK only really provides one of the three the i686 (i.e. Mingw32) one.

jfmherokiller commented:
> 
> 
> > btw we can check for msys2 by using if(MSYS) inside of cmake i think
> 
> You mean `if(MSYS2)` don't you?
> 
> We have to be a bit cautious as there are some bits that it shares with the project from which it was forked and Qt's own Windows SDK does use some elements that are shared with the MSYS2/Mingw-w64 environment.
> 
> Effectively the full MSYS2 setup actually provides three separate environments a bare-bones more *nix-like "MSYS2" one which is where auto-tool-ised libraries can be built from the command line and two more Windows like "Mingw32" and "Mingw64" environments where GUI applications are built for the corresponding bit-nesses. The Qt SDK only really provides one of the three the i686 (i.e. Mingw32) one.

ah well in cmake to handle x86 and x64 i use this construct

```
if(CMAKE_SIZEOF_VOID_P EQUAL 8)
    set(VC_LIB_PATH_SUFFIX lib/x64)
else()
    set(VC_LIB_PATH_SUFFIX lib/x86)
endif()
```

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -17158,49 +17157,90 @@ void TLuaInterpreter::initIndenterGlobals()
+    // NOTE: In the cmake case the shadow build directory is alongside the base
+    // directory of the extracted source package so the source code MUST be
+    // extracted into a known directory name which is assumed here to be
+    // "mudlet"!
```
Having this information here is the equivalent of hiding it really, really well inside 17,000 lines! This isn't the right place for it: at most, it should be added to the compiling instructions for cmake.

SlySven replied:
Okay - I guess... :neutral_face:

vadi2 commented on src/mudlet.pro:
```diff
@@ -271,23 +281,51 @@ unix:!macx {
+        # For users/developers building with MSYS2 on Windows:
+        isEmpty( MINGW_BASE_DIR_TEST ) {
+            # As error(...) terminates the build process immediately a
+            # multi-line message must be presented as a single string!
```
We don't need an explanation of how the error() function works here.

SlySven replied:
Well, it was the first one in that is more than one line - and the temptation to break the message up into several `error(...)` lines was there for me until I realised that that wouldn't work. Guess it can go though - lets see if anyone in the future falls into the trap though I guess they won't with this one to copy from - I must confess I wasn't aware of the `$$escape_expand(...)` function before this! :open_mouth:

SlySven replied:
Done - well more accurately *undone*.

keneanung commented on src/CMakeLists.txt:
```diff
@@ -357,31 +357,56 @@ target_compile_definitions(mudlet
-  ZLIB::ZLIB
-  qt5keychain)
+if(WIN32 AND NOT USE_MAIN_BUILD_SYSTEM)
+  # For this combination the use of an INTERFACE definition (with the library
```
This might sound a bit hard, but I have multiple issues with this:
1. Why the scope creep? Why not make MSYS2 working here with qmake (the build system we use on windows) Then add CMake changes in a follow-up PR?
2. Not using the interface definition seems to be working around insufficient find module definitions. Why don't we fix those instead of doing something half baked here?
3. If we must do the workaround, why not extract the differentiating libraries into a block, leaving the common code common?

SlySven replied:
> 1. Why the scope creep? Why not make MSYS2 working here with qmake (the build system we use on windows) Then add CMake changes in a follow-up PR?

There is a separate thing that we want to also get into the project (a crash-handler) but that only comes with CMake build files at the moment - I am working on getting some QMake ones in upstream but I don't want to rely on that aspect going in straight away. Also, Windows is not my primary OS - and it is more convenient to me to work on both build systems whilst I am mentally in Windoze mode - and toggling between them helps to ensure that I am matching up the details/requirements for both equally - and allows me to compare the results so that they both work.

> 2. Not using the interface definition seems to be working around insufficient find module definitions. Why don't we fix those instead of doing something half baked here?

If you understand WTF is going on with those I would welcome your input - all I can say is that NOT using the `LIBRARYNAME::LIBRARYNAME` interface form for those libraries - which is what CMake seems to call it - fixes a bloody peculiar error - and that the CMake documentation is short on explanations or samples that actually show what is going on.

> 3. If we must do the workaround, why not extract the differentiating libraries into a block, leaving the common code common?

That is probably fair comment - I wasn't entirely sure that the order of listing was important initially - but now I realise it isn't so yes that can be made a smaller difference.

SlySven replied:
2. :bulb: Actually, with regard to this point, since in the MSYS2/Mingw-w64 Windows platform most, if not all, the libraries involved ARE effectively system libraries as far as the compiler/linker toolchain is concerned (i.e. they come as part of the system) and in that situation we are not building the libraries from source code ourselves. So they must not have the interface type declaration (name repeated with a `::` between those) - instead they are IMPORTED libraries so are specified by their name in a singular form.

3. Done.

keneanung replied:
> So they must not have the interface type declaration (name repeated with a :: between those) - instead they are IMPORTED libraries so are specified by their name in a singular form.

This is wrong. The `::` definitions are `IMPORTED` libraries as they should be. My guess is that the way you added the libraries, cmake is adding simply `-l hunspell` as linker flags. Which might work purely by chance, because the libraries are in the search path as expected. Transitive dependencies, linker flags or compile definitions are NOT honoured this way.

I may be able to help you get this right, but I'll need a clear way how to set up a MSYS2 compile environment for Mudlet.

SlySven replied:
Ah, diving deeper down this rabbit hole it seems to me that both the IMPORTED and INTERFACE methods are for libraries that are built-with or at least accompany the main target in some way. However the libraries being separated out in this way are, when used in an MSYS2/Mingw-w64 Mudlet build, *normal* system-like libraries whereas if the original Windows build method is used they {the `hunspell`, `pcre`, `yajl` and `zip` libraries) are also built as part of the process.

keneanung replied:
No. https://cmake.org/cmake/help/v3.18/command/add_library.html#imported-libraries explicitly states:
> An IMPORTED library target references a library file located outside the project. No rules are generated to build it

SlySven replied:
🤷‍♂️ Okay, I am wrong in my theorising about what is happening. However whatever is the situation, the method I am using does *seem* to work. If it can be improved to be more like the other cases then I am not opposed to that - but so far using the LIBRARYNAME::LIBRARYNAME form for those libraries just breaks things in ways that the CMake documentation doesn't seem to cover...

keneanung replied:
The underlying issue is now fixed with commit 6a95a41e1808ae51015986ade52a4bdaa7d8bad2.

When using the libraries in question, we had to differentiate between static and shared libraries. The static versions of these libraries required additional compile definitions, the shared versions did not.

Apparently, the detection, whether static or shared versions were used, was broken. This lead to the assumption of cmake, that the libraries were not identified correctly. This, in turn, made cmake add `LIBRARY::LIBRARY-NOTFOUND` to the Makefile, on which `make` choked.

Debugging was rather easy since the error message explicitly stated the line number on which `make` stopped. A few debug `MESSAGE` statements quickly identified the issue.

mpconley commented:
The lua code formatter worked for me.  I did not get any unanticipated errors upon startup running on MacOS.  I did not compile from source, just ran the MacOS version so far.

SlySven committed:
Revise: implement some changes requested by peer-review

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into windows_development

Needed to update this PR.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: shorten multi-line comment in initIndenterGlobals()

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
Just to reiterate - the point of this PR is to make it simpler to setup a coherent Windows SDK to build both 32 and 64 bit Mudlet applications.

Whilst the existing setup PS script might be fine for some, others (like myself) find the MSYS2/Mingw-w64 SDK a more robust environment as it is provides the libraries and tools all built with the same toolchains and which handles both bit-nesses - which the current Windows offering cannot (IMHO).

Also, it is relatively painless to produce builds that can be debugged interactively which is something that is, I think, problematic with the existing system.

I think that this PR should be enough to enable myself and other interested parties to make contributions to the Mudlet project from Windows - and until it goes in it is a right faff for me to investigate other Windows specific issues because I have to resort to cherry-picking commits for those issues on-top of a branch like this instead of the mainline `development` one! :scream: 

FTR - to use this all that should be needed is to define the following two environmental variables (which I stick in the relevant section of the project build page in Qt Creator) to something like:
```sh
MINGW_BASE_DIR=C:/msys64/mingw64
WITH_MAIN_BUILD_SYSTEM=NO
```
far enough it may **also** be desirable/needed to do some clean-up of the `PATH` variable that is used before calling `qtcreator &` from ~~a `MING32`~~ (*Edit: that is a GUI process so must be run from a MINGW64 or MINGW32 environment*) ~~or~~ `MINGW64` shell and to also create/revises the `LUA_PATH`and `LUA_CPATH` environmental variables to include the user specific luarocks directories but those are not insurmountable issues.

vadi2 reviewed (commented):
Thanks! Looking alright with me now - will leave it up to demonnic and keneanung to review this as well.

SlySven commented:
@vadi2 it looks like I will not be able to merge until you and @keneanung are happy that the changes you have both requested are done. If you both (or perhaps just he) can refresh me on the sticking points maybe we can work them out... :pray:

vadi2 commented:
I'm good, just @keneanung is outstanding

keneanung commented on src/CMakeLists.txt:
```diff
@@ -407,6 +421,16 @@ if(USE_VARIABLE_SPLASH_SCREEN)
   target_compile_definitions(mudlet PRIVATE INCLUDE_VARIABLE_SPLASH_SCREEN)
 endif()
 
+if(WIN32)
```
Why is this block extra and not combined with the `WIN32` block below?

SlySven replied:
Done.

keneanung commented on src/CMakeLists.txt:
```diff
@@ -407,6 +421,16 @@ if(USE_VARIABLE_SPLASH_SCREEN)
+    # (QString) mudlet::getShortPathName(const QString& name) work:
+    target_compile_definitions(mudlet PRIVATE INCLUDE_MAIN_BUILD_SYSTEM UNICODE _UNICODE)
+  else()
+    target_compile_definitions(mudlet PRIVATE UNICODE _UNICODE)
```
This can be outside the `MAIN_BUILD_SYSTEM`-if. Repeated calls to the function append compile definitions.

SlySven replied:
Will do...

SlySven replied:
... done.

keneanung reviewed (changes requested):
Still not happy with some of the CMake changes. I also added https://github.com/Mudlet/Mudlet/pull/3889/files#r475469677 but hit the wrong button :see_no_evil:

SlySven committed:
Revise: further change suggested in peer-review

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

keneanung committed:
Fix recognition of static libraries in own Find modules

keneanung reviewed (approved):
I'm happy now with the cmake issue resolved (though others might want to cast a second pair of eyeballs over my changes).

vadi2 commented:
@SlySven can 64bit builds be enabled in appveyor now?

SlySven commented:
Nope - not yet - this just does the changes needed to build something locally on Windows. I need to revisit #3255 and review the part (that I left out of this one) that builds an archive (zip) file containing the `mudlet.exe` file and all the MSYS2 / Mingw-w64/ Qt / other libraries (and our Lua support files) that can be plonked into a Windows (either a 32 or a 64 bit one) PC ***without*** an MSYS2 / Mingw-w64 installation but still run.

The system I had cobbled together were some `bash` scripts (I really detest PowerShell and I could do everything without it) - though I had only got as far as that single archive (which I arranged to be placed in a separate `./package` sub-directory) - I had not decoded the mechanism that converted that into an installable product. Also, I had left in the `WITH_UPDATER` environmental variable set to `NO` because it is not clear how to set up a separate 64-Bit as well as a 32-Bit updater channel for Windows.

SlySven commented:
BTW thanks @keneanung for working out the `FindXXXX` stuff - it really went over my head. Sadly, I thought it might give me a hint how to get CMake builds working on FreeBSD - but that is still failing for me in the `QtKeyChain` additional libraries needing to be linked to that - both with the bundled one AND the system one (that FreeBSD has) - the failed linking step for the latter ending:
```sh
[ 85%] Linking CXX executable mudlet
cd /home/stephen/src/mudlet/build-mudlet-Desktop_Qt_5_13_2_CMake_Clang_9_0_0-Debug/src && /usr/local/bin/cmake -E cmake_link_script CMakeFiles/mudlet.dir/link.txt --verbose=1
ccache /usr/local/llvm90/bin/clang-9  -g   CMakeFiles/mudlet.dir/mudlet_autogen/mocs_compilation.cpp.o CMakeFiles/mudlet.dir/ActionUnit.cpp.o CMakeFiles/mudlet.dir/AliasUnit.cpp.o CMakeFiles/mudlet.dir/ctelnet.cpp.o CMakeFiles/mudlet.dir/discord.cpp.o CMakeFiles/mudlet.dir/dlgAboutDialog.cpp.o CMakeFiles/mudlet.dir/dlgActionMainArea.cpp.o CMakeFiles/mudlet.dir/dlgAliasMainArea.cpp.o CMakeFiles/mudlet.dir/dlgColorTrigger.cpp.o CMakeFiles/mudlet.dir/dlgComposer.cpp.o CMakeFiles/mudlet.dir/dlgConnectionProfiles.cpp.o CMakeFiles/mudlet.dir/dlgIRC.cpp.o CMakeFiles/mudlet.dir/dlgKeysMainArea.cpp.o CMakeFiles/mudlet.dir/dlgMapper.cpp.o CMakeFiles/mudlet.dir/dlgNotepad.cpp.o CMakeFiles/mudlet.dir/dlgPackageExporter.cpp.o CMakeFiles/mudlet.dir/dlgProfilePreferences.cpp.o CMakeFiles/mudlet.dir/dlgRoomExits.cpp.o CMakeFiles/mudlet.dir/dlgScriptsMainArea.cpp.o CMakeFiles/mudlet.dir/dlgSourceEditorArea.cpp.o CMakeFiles/mudlet.dir/dlgSourceEditorFindArea.cpp.o CMakeFiles/mudlet.dir/dlgSystemMessageArea.cpp.o CMakeFiles/mudlet.dir/dlgTimersMainArea.cpp.o CMakeFiles/mudlet.dir/dlgTriggerEditor.cpp.o CMakeFiles/mudlet.dir/dlgTriggerPatternEdit.cpp.o CMakeFiles/mudlet.dir/dlgTriggersMainArea.cpp.o CMakeFiles/mudlet.dir/dlgVarsMainArea.cpp.o CMakeFiles/mudlet.dir/EAction.cpp.o CMakeFiles/mudlet.dir/exitstreewidget.cpp.o CMakeFiles/mudlet.dir/FontManager.cpp.o CMakeFiles/mudlet.dir/Host.cpp.o CMakeFiles/mudlet.dir/HostManager.cpp.o CMakeFiles/mudlet.dir/ircmessageformatter.cpp.o CMakeFiles/mudlet.dir/KeyUnit.cpp.o CMakeFiles/mudlet.dir/LuaInterface.cpp.o CMakeFiles/mudlet.dir/main.cpp.o CMakeFiles/mudlet.dir/mudlet.cpp.o CMakeFiles/mudlet.dir/ScriptUnit.cpp.o CMakeFiles/mudlet.dir/T2DMap.cpp.o CMakeFiles/mudlet.dir/TAction.cpp.o CMakeFiles/mudlet.dir/TAlias.cpp.o CMakeFiles/mudlet.dir/TArea.cpp.o CMakeFiles/mudlet.dir/TBuffer.cpp.o CMakeFiles/mudlet.dir/TCommandLine.cpp.o CMakeFiles/mudlet.dir/TConsole.cpp.o CMakeFiles/mudlet.dir/TDebug.cpp.o CMakeFiles/mudlet.dir/TDockWidget.cpp.o CMakeFiles/mudlet.dir/TEasyButtonBar.cpp.o CMakeFiles/mudlet.dir/TEncodingTable.cpp.o CMakeFiles/mudlet.dir/TEntityHandler.cpp.o CMakeFiles/mudlet.dir/TEntityResolver.cpp.o CMakeFiles/mudlet.dir/TFlipButton.cpp.o CMakeFiles/mudlet.dir/TForkedProcess.cpp.o CMakeFiles/mudlet.dir/TimerUnit.cpp.o CMakeFiles/mudlet.dir/TKey.cpp.o CMakeFiles/mudlet.dir/TLabel.cpp.o CMakeFiles/mudlet.dir/TLinkStore.cpp.o CMakeFiles/mudlet.dir/TLuaInterpreter.cpp.o CMakeFiles/mudlet.dir/TMap.cpp.o CMakeFiles/mudlet.dir/TMedia.cpp.o CMakeFiles/mudlet.dir/TMxpElementDefinitionHandler.cpp.o CMakeFiles/mudlet.dir/TMxpElementRegistry.cpp.o CMakeFiles/mudlet.dir/TMxpFormattingTagsHandler.cpp.o CMakeFiles/mudlet.dir/TMxpBRTagHandler.cpp.o CMakeFiles/mudlet.dir/TMxpColorTagHandler.cpp.o CMakeFiles/mudlet.dir/TMxpCustomElementTagHandler.cpp.o CMakeFiles/mudlet.dir/TMxpEntityTagHandler.cpp.o CMakeFiles/mudlet.dir/TMxpFontTagHandler.cpp.o CMakeFiles/mudlet.dir/TMxpLinkTagHandler.cpp.o CMakeFiles/mudlet.dir/TMxpNodeBuilder.cpp.o CMakeFiles/mudlet.dir/TMxpMudlet.cpp.o CMakeFiles/mudlet.dir/TMxpProcessor.cpp.o CMakeFiles/mudlet.dir/TMxpSendTagHandler.cpp.o CMakeFiles/mudlet.dir/TMxpSupportTagHandler.cpp.o CMakeFiles/mudlet.dir/MxpTag.cpp.o CMakeFiles/mudlet.dir/TMxpTagHandler.cpp.o CMakeFiles/mudlet.dir/TMxpTagParser.cpp.o CMakeFiles/mudlet.dir/TMxpTagProcessor.cpp.o CMakeFiles/mudlet.dir/TMxpVarTagHandler.cpp.o CMakeFiles/mudlet.dir/TMxpVersionTagHandler.cpp.o CMakeFiles/mudlet.dir/TriggerUnit.cpp.o CMakeFiles/mudlet.dir/TRoom.cpp.o CMakeFiles/mudlet.dir/TRoomDB.cpp.o CMakeFiles/mudlet.dir/TScript.cpp.o CMakeFiles/mudlet.dir/TSplitter.cpp.o CMakeFiles/mudlet.dir/TSplitterHandle.cpp.o CMakeFiles/mudlet.dir/TStringUtils.cpp.o CMakeFiles/mudlet.dir/TTabBar.cpp.o CMakeFiles/mudlet.dir/TTextCodec.cpp.o CMakeFiles/mudlet.dir/TTextEdit.cpp.o CMakeFiles/mudlet.dir/TTimer.cpp.o CMakeFiles/mudlet.dir/TToolBar.cpp.o CMakeFiles/mudlet.dir/TTreeWidget.cpp.o CMakeFiles/mudlet.dir/TTrigger.cpp.o CMakeFiles/mudlet.dir/TVar.cpp.o CMakeFiles/mudlet.dir/VarUnit.cpp.o CMakeFiles/mudlet.dir/XMLexport.cpp.o CMakeFiles/mudlet.dir/XMLimport.cpp.o CMakeFiles/mudlet.dir/glwidget.cpp.o CMakeFiles/mudlet.dir/mudlet_autogen/EWIEGA46WW/qrc_mudlet.cpp.o  -o mudlet  -Wl,-rpath,/usr/local/lib:/usr/local/lib/qt5:/home/stephen/src/mudlet/build-mudlet-Desktop_Qt_5_13_2_CMake_Clang_9_0_0-Debug/3rdparty/edbee-lib/edbee-lib/qslog/lib: ../3rdparty/communi/libcommuni.a ../3rdparty/edbee-lib/edbee-lib/libedbee-lib.a /usr/local/lib/libhunspell-1.7.so /usr/local/lib/liblua-5.1.so /usr/local/lib/libpcre.so /usr/local/lib/libpugixml.so /usr/local/lib/qt5/libQt5Concurrent.so.5.14.2 /usr/local/lib/qt5/libQt5Multimedia.so.5.14.2 /usr/local/lib/qt5/libQt5OpenGL.so.5.14.2 /usr/local/lib/qt5/libQt5UiTools.a /usr/local/lib/qt5/libQt5Widgets.so.5.14.2 /usr/local/lib/libyajl.so /usr/local/lib/libzip.so /usr/lib/libz.so /usr/local/lib/libqt5keychain.so.0.10.0 /usr/local/lib/qt5/libQt5TextToSpeech.so.5.14.2 /usr/local/lib/libGLU.so /usr/local/lib/qt5/libQt5Network.so.5.14.2 /usr/local/lib/qt5/libQt5Core.so.5.14.2 ../3rdparty/edbee-lib/edbee-lib/qslog/lib/libQsLog_d.so /usr/local/lib/qt5/libQt5Core.so.5.14.2 /usr/lib/libm.so /usr/lib/libdl.so /usr/local/lib/qt5/libQt5Network.so.5.14.2 /usr/local/lib/qt5/libQt5Gui.so.5.14.2 /usr/local/lib/qt5//libQt5Widgets.so /usr/local/lib/qt5//libQt5Gui.so /usr/local/lib/qt5//libQt5Core.so /usr/local/lib/qt5//libQt5Widgets.so /usr/local/lib/qt5//libQt5Gui.so /usr/local/lib/qt5//libQt5Core.so /usr/local/lib/libGL.so -lsecret-1 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lintl /usr/local/lib/qt5/libQt5DBus.so.5.14.2 /usr/local/lib/qt5/libQt5Core.so.5.14.2 /usr/local/lib/libGL.so -Wl,-rpath-link,/usr/local/lib 
ld: error: unable to find library -lsecret-1
ld: error: unable to find library -lgio-2.0
ld: error: unable to find library -lgobject-2.0
ld: error: unable to find library -lglib-2.0
ld: error: unable to find library -lintl
clang-9: error: linker command failed with exit code 1 (use -v to see invocation)
gmake[2]: *** [src/CMakeFiles/mudlet.dir/build.make:1833: src/mudlet] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:425: src/CMakeFiles/mudlet.dir/all] Error 2
gmake: *** [Makefile:164: all] Error 2
gmake[2]: Leaving directory '/home/stephen/src/mudlet/build-mudlet-Desktop_Qt_5_13_2_CMake_Clang_9_0_0-Debug'
gmake[1]: Leaving directory '/home/stephen/src/mudlet/build-mudlet-Desktop_Qt_5_13_2_CMake_Clang_9_0_0-Debug'
05:15:36: The process "/usr/local/bin/cmake" exited with code 2.
Error while building/deploying project mudlet (kit: Desktop Qt 5.14.2 Clang-9.0.0)
When executing step "CMake Build"
```

What might not be obvious is that the tail end does list those missing libs but only in the `-l` manner:
`-lsecret-1 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lintl` whereas other libraries are given as a full absolute pathname and not as an `-l` option.

Not sure either about the errors being reported by the `ld` executable given that the advanced CMake details show that the `CMAKE_LINKER` is, as is appropriate on FreeBSD, the `llvm-link90` (i.e. the version 9.0 `llvm-link` as both llvm 8 and 9 are installed) one and not the GCC linker...!
